message/history/eventFinder.go

```diff
@@ -0,0 +1,182 @@
+			latestEvent = compareEvents(event, latestEvent, validators, prevBootTime)
+		}
+
+		// final check to make sure that we actually found an event
+		return checkEvent(latestEvent, prevBootTime)
+	}
+
+}
+
+// CurrentSessionFinder returns a function to find an event that is deemed valid by the Validator passed in
+// with the boot-time of the current event. If any of the fatalValidators returns false,
+// it will stop searching and immediately exit, returning the event
+// that prompted the error along with the error returned by the fatalValidator.
+func CurrentSessionFinder(validators validation.Validator, fatalValidators validation.Validator) FinderFunc {
+	return func(events []message.Event, currentEvent message.Event) (message.Event, error) {
+		// verify that the current event has a boot-time
+		currentBootTime, err := currentEvent.BootTime()
+		if currentBootTime <= 0 {
+			return message.Event{}, validation.InvalidBootTimeErr{OriginalErr: err}
+		}
+
+		var latestEvent message.Event
+		for _, event := range events {
+			// if transaction UUIDs are the same, continue onto next event
+			if event.TransactionUUID == currentEvent.TransactionUUID {
+				continue
+			}
+
+			// if any fatalValidators return false, it means we should stop looking for an event
+			// because there is something wrong with currentEvent, and we should not
+			// perform calculations using it.
+			if valid, err := fatalValidators.Valid(event); !valid {
+				return event, validation.InvalidEventErr{OriginalErr: err}
+			}
+
+			// Get the bootTime from the event we are checking. If boot-time
+			// doesn't exist, move on to the next event.
+			bootTime, _ := event.BootTime()
+			if bootTime <= 0 {
+				continue
+			}
+
+			latestEvent = compareEvents(event, latestEvent, validators, currentBootTime)
+		}
+
+		// final check to make sure that we actually found an event
+		return checkEvent(latestEvent, currentBootTime)
+	}
+}
+
+// SameEventFinder returns a function that goes through a list of events and compares the currentEvent
+// to these events to make sure that currentEvent is valid. If any of the fatalValidators returns false,
+// it will stop searching and immediately exit, returning the event
+// that prompted the error along with the error returned by the fatalValidator. If all of the fatalValidators
+// pass, the currentEvent is returned along with nil error.
+func SameEventFinder(fatalValidators validation.Validator) FinderFunc {
+	return func(events []message.Event, currentEvent message.Event) (message.Event, error) {
+		// verify that the current event has a boot-time
+		currentBootTime, err := currentEvent.BootTime()
+		if currentBootTime <= 0 {
+			return message.Event{}, validation.InvalidBootTimeErr{OriginalErr: err}
+		}
+
+		for _, event := range events {
+			// if transaction UUIDs are the same, continue onto next event
+			if event.TransactionUUID == currentEvent.TransactionUUID {
+				continue
```

Comment:
The name of this function focuses on finding the same event in the list, but it seems like the main goal of this function is actually to validate every event in the list? Otherwise, I'm not sure why we don't just return after finding the event we're looking for?

Reply:
So the goal of this function (and of the goals of `LastSessionFinder` and `CurrentSessionFinder`) is to make sure that the incoming event from caduceus is valid compared to all of the events in the device's history. This is to make sure that the incoming event doesn't have an old boot-time or that a duplicate of the event wasn't already received and processed by glaukos. I guess this might be outside the scope of a `Finder` but I didn't want to loop through events twice and thought I'd do the caduceus event validation as I looped through to find the proper codex event.